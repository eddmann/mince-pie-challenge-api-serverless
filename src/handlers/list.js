// @flow

import { createHandler } from '../helpers/handlers';
import { ok } from '../helpers/http';
import { Resource } from 'hal';

const list = async ({ services: { allPies } }) => {
  const pies = (await allPies()).map(
    pie =>
      new Resource(
        {
          id: pie.Id,
          name: pie.Name,
          thumbnail: pie.ThumbnailUrl,
          rating: { avg: pie.AvgRating, total: pie.TotalRatings },
        },
        `/pies/${pie.Id}`,
      ),
  );

  const resource = new Resource({ total: pies.length }, '/pies');

  resource.embed('pies', pies);

  return ok(resource);
};

export default createHandler(list);
