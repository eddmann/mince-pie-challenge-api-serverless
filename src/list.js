import * as db from './helpers/db';
import { authenticate } from './middleware/auth';
import { ok } from './helpers/http';
import { Resource } from 'hal';

export const list = authenticate(async function (event, context, callback, user) {
  const pies = (await db.all()).map(pie => new Resource({
    id: pie.Id,
    name: pie.Name,
    thumbnail: pie.ThumbnailUrl,
    rating: { avg: pie.AvgRating, total: Object.keys(pie.Ratings).length },
  }, `/pies/${pie.Id}`));

  const resource = new Resource({
    total: pies.length,
  }, '/pies');

  resource.embed('pies', pies);

  callback(undefined, ok(resource));
});
