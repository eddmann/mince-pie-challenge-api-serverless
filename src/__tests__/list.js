import list from '../handlers/list';

const response = () =>
  function cb(error, response) {
    if (error) fail(`Unexpected error: ${error}`);
    cb.response = response;
  };

it('list mince pies', async () => {
  const pies = [
    {
      Id: '1',
      UserId: 'USER_ID',
      Name: 'Mince Pie',
      AvgRating: 0,
      PhotoUrl: undefined,
      ThumbnailUrl: undefined,
      Ratings: {},
      AddedAt: '123456789',
    },
    {
      Id: '2',
      UserId: 'USER_ID',
      Name: 'Another Mince Pie',
      AvgRating: 0,
      PhotoUrl: undefined,
      ThumbnailUrl: undefined,
      Ratings: {},
      AddedAt: '123456789',
    },
  ];

  const services = {
    db: { all: () => Promise.resolve(pies) },
  };

  const callback = response();

  await list(services)({}, {}, callback, undefined);

  expect(callback.response).toMatchSnapshot();
});
