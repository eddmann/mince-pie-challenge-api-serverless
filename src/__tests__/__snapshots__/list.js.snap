// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list mince pies 1`] = `
Object {
  "body": "{\\"_links\\":{\\"self\\":{\\"href\\":\\"/pies\\"}},\\"_embedded\\":{\\"pies\\":[{\\"_links\\":{\\"self\\":{\\"href\\":\\"/pies/1\\"}},\\"id\\":\\"1\\",\\"name\\":\\"Mince Pie\\",\\"rating\\":{\\"avg\\":0,\\"total\\":0}},{\\"_links\\":{\\"self\\":{\\"href\\":\\"/pies/2\\"}},\\"id\\":\\"2\\",\\"name\\":\\"Another Mince Pie\\",\\"rating\\":{\\"avg\\":0,\\"total\\":0}}]},\\"total\\":2}",
  "headers": Object {
    "Access-Control-Allow-Credentials": "*",
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/hal+json",
  },
  "statusCode": 200,
}
`;
